### ISSUE TITLE
Can we handle connections among more than two edges?

### ISSUE THREAD

**Issue by refraction-ray** [google/TensorNetwork#552]
Something like "im,jm,km->ijk", or in TN graph 
```bash
- o -   - o -
       |
       o
       |
```

This is a legal and reasonable einsum expression,  is such thing possible in current setups with TensorNetwork?

**Comment by Thenerdstation** [comment]
Yes but indirectly. It's called a `CopyNode`.

```python
nodes =[...]
center = tn.CopyNode(...)
for i, node in enumerate(nodes):
  tn.connect(center[i], node[0])
result = tn.contract_copy_node(center)
```

**Comment by Thenerdstation** [comment]
But this is I believe the 3rd time time someone had requested this. We really should consider supporting these hyper edges.

**Comment by refraction-ray** [comment]
> Yes but indirectly. It's called a CopyNode.

Thanks, I will have a look at CopyNode approach.

> But this is I believe the 3rd time time someone had requested this. We really should consider supporting these hyper edges.

It would be great if such connection can be supported in a more natural way. But I think the main issue here is the lack of good documentations and tutorials. For example, there is no explanation on CopyNode even in docstrings. Currently one has to read the source to figure out what is going on for nearly everything. We need more documentations to cover basics as well as contractor, MPS, and quantum subpackages.

**Comment by Thenerdstation** [comment]
That we very much do. The main issue is that this package is still very new and many of the APIs we have are still changing (we haven't felt comfortable to release a 1.0 yet).

Though you're right, the basic tutorial we have isn't nearly enough for things that are stable. I'll update with some better tutorials tomorrow night.

**Comment by MichaelMarien** [comment]
If the request keeps coming up, I guess I was wrong about not wanting #401 , is that one still our preferred option?

**Comment by Thenerdstation** [comment]
No it really isn't. In fact there are many cases where doing the contraction like above is more expensive than just contracting the CopyNode like a regular node.

The reason we haven't done this is because I made a mistake really early in development, and now all of the code has a REALLY strong dependence on the edges being normal edges (and not hyper edges). You can see this in the code base with all of the `edge.node1` and `edge.node2` usage.

Let's reopen #401

**Comment by MichaelMarien** [comment]
That not a mistake man, that's good old YAGNI.

Do you think we're just adding technical debt by not addressing it now in the best way possible? Even if it is a lot of work to change it, it's never gonna be less work than now. If we decide that it's a top priority, plan it well and divide work over 2-3 of us, it might not be that bad and give us a lot more flexibility?

**Comment by Thenerdstation** [comment]
Yeah you're right, we should fix it now.

The biggest thing is we need to do is remove the attributes `axis1`, `axis2`, `node1`, `node2` on the `Edge` class and replace them with lists `axes` and `nodes`. We also will need to add a `batched_tensordot` operation on all of the backends. Beyond that, the rest is just patching for all of the methods to accommodate this change.

To make the transition a little easier, we can set the above attributes as `@property` methods, while having the underlying function just work with the lists. Once all of the methods are transitioned away, we can deleted the `@property` methods.

**Comment by Thenerdstation** [comment]
Some tutorials have been added around copying/contracting and a simple tutorial around building a TN for quantum circuits. Still not enough, but it's a little better.

I think ultimately we should remove `CopyNode` in favor of hyper edges, though I'm not sure about the timeline.

**Comment by alewis** [comment]
It's maybe worth pointing out that although np.einsum lets your write it, "im,jm,km->ijk" is actually ill-formed Einstein notation. While since our "tensors" have no metric it's probably harmless, the desired operation would in fact be written by explicitly including a rank-3 identity tensor, with which `CopyNode` seems closer in spirit.

**Comment by mganahl** [comment]
what is our current strategy regarding `CopyNode` like contractions?

In general, such things could be handled using batched contractions followed by summation. The former would be very nice to have as well for `ncon`.

**Comment by alewis** [comment]
Even though reading it fills me with pain, supporting this notation in ncon is probably simplifying enough to be worthwhile.

**Comment by Thenerdstation** [comment]
The main problem is that copy tensors are fairly inefficient (See #261).

Really, I think the answer is to just add support for hyper edges #401.

**Comment by mganahl** [comment]
yes, the current `CopyNode` behaviour is unsatisfactory, and I believe we should remove it completely.

From a conceptual viewpoint such things would be very easy to implement once we have a tensordot function that allows batching. 

I have now encountered such cases already several times in my own research, and would find it very convenient to have it. In my case, I would favour an `ncon` which supports batching

**Comment by alewis** [comment]
I'm a little nervous about encouraging users to draw invalid tensor
networks. It seems to me that, at least conceptually, if this behaviour is
supported it should be documented and explained as implicitly introducing a
Kronecker delta "tensor" at the junction (e.g. the fusing is accomplished
in the conceptual tensor network by a node instead of an edge; this doesn't
however speak to whether or not the implementation should involve some kind
of Edge behind the scenes). For example, the contraction order becomes
ambiguous otherwise.

It's a little unclear to me what actual floating point operations would
occur here. Would e.g. A_{ab} B_{bc} C_{bd} still reduce to a series of
matrix multiplications, or is the idea to do "the Fortran thing", e.g.
for a..
   for b...
     for c...
       for d...
          result[a,c,d] += A[a,b]*B[b,c]*C[b,d]

It's not obvious to me how this gets written as batched matrix
multiplications (I believe you that it can be though).

Another concern is that this is somewhat orthogonal to our present goals
w.r.t. getting PEPS working and so on.

**Comment by mganahl** [comment]
That's right, and I'm not suggesting to do it right away, but we should keep it on the agenda. If this is ever going to be a thing for machine learning then batched contractions are a must. 
A frequent case is where one has a tensor network where each tensor has a batch index:
```python 
A = np.random.rand(batchsize,D1,D2)
B = np.random.rand(batchsize,D2,D3)
C = np.random.rand(batchsize,D3,D1)
```
and one wants to contract over `D1, D2, D3` and sum over `batchsize`. Typically this is done
using `np.matmul` (or `tf.matmul`):

```python 
AB = np.matmul(A,B) #shape (batchsize, D1, D3)
ABC = np.matmul(AB,C) #shape (batchsize, D1,D1)
cost = np.sum(np.trace(ABC))
```
It's easy to extend this to cases where using `np.matmul` will be a real pain. `ncon` could handle
this e.g. by using a `batchindex` argument

```python 
cost = ncon([A,B,C],[[4,1,2],[4,2,3],[4,3,1]], batchindex=[4])
```
Also note that not all tensors neccessarily need to have a batch index, e.g.
```python 
A = np.random.rand(batchsize,D1,D2)
B = np.random.rand(D2,D3)
C = np.random.rand(batchsize,D3,D1)
cost = ncon([A,B,C],[[4,1,2],[2,3],[4,3,1]], batchindex=[4])
```
should also be possible.

**Comment by alewis** [comment]
I see. Well that makes sense to me, and the meaning seems clear enough. I
do like making the user repeat themself with "batchindex" to prove the y
didn't just make a mistake. Is this the same thing as a "hyper-edge"?

**Comment by mganahl** [comment]
The hyper edge would connect to multiple tensors and do the batch summation. 
The `ncon` API seems more convenient . Consider this case
```python 
A = np.random.rand(batchsize,D1,D2)
B = np.random.rand(D2,D3)
C = np.random.rand(batchsize,D3,D1)
cost = ncon([A,B,C],[[-1,1,2],[2,3],[-1,3,1]], batchindex=[-1]) #shape (batchsize,)
```
One could in this case connect `A` and `C` with a hyper edge, but not contract it. 
While there is no fundamental problem with it, I don't quite like that the API defines an object that it doesn't really use in the end.

**Comment by Thenerdstation** [comment]
There would be not need to specify a batchindex, and negative number that appears more than once would be treated as a batch index.

Of course, you may choose to include for error catch, but you don't NEED to have it.

**Comment by Thenerdstation** [comment]
@alewis, your comment on how do you contract A_ab, B_bc and C_bd with b being a hyper edge looks like this:

First, you do a batched outerproduct of A and B together to get

D_bac = A_ab @ B_bc

Next, you do a normal tensordot operation of D and C

Result_acd = D_bac @ C_bd

Rule of thumb: If an index is on more than 2 nodes of the network at that instance of contraction, you treat that index as a batch index and the reset as you normally would.

**Comment by alewis** [comment]
Indeed, it is redundant apart from error catch. I think physicists vs deep
learning types are likely to have opposite instincts wrt to whether or not
this behaviour is sufficiently natural to require said catch, so it's not
clear to me which is best.

**Comment by mganahl** [comment]
> There would be not need to specify a batchindex, and negative number that appears more than once would be treated as a batch index.
> 
> Of course, you may choose to include it for error catching, but you don't NEED to have it.

For batch-contracted indices one still needs to specify it. For uncontracted ones one could leave it, but this would lead to confusing behaviour if the user accidentally puts two identical negative numbers when he meant to use two different ones.

**Comment by Thenerdstation** [comment]
Wouldn't a batch-contracted index just be a negative number that appears 3 times? Or am I misunderstanding this?

Can you give an example martin.

**Comment by mganahl** [comment]
actually you're right, one could leave it

**Comment by Thenerdstation** [comment]
Ultimately, ncon has now come full circle to just `np.einsum`. :)

**Comment by Thenerdstation** [comment]
Though ncon gives you direct control of the contraction order, so that's still useful.
